drv/i2c-devices/src/emc2305.rs

```diff
@@ -0,0 +1,334 @@
+    device.write(&[register as u8, val])
+}
+
+fn write_reg16(
+    device: &I2cDevice,
+    register: Register,
+    val: u16,
+) -> Result<(), ResponseCode> {
+    device.write(&[register as u8, (val >> 8) as u8, (val & 0xff) as u8])
+}
+
+#[derive(Copy, Clone, PartialEq)]
+enum Trace {
+    ZeroTach(Fan),
+    TachOverflow(u32),
+    BadFanCount(u8),
+    None,
+}
+
+ringbuf!(Trace, 6, Trace::None);
+
+impl Emc2305 {
+    pub fn new(device: &I2cDevice) -> Self {
+        Self { device: *device }
+    }
+
+    pub fn initialize(&self, fan_count: u8) -> Result<(), ResponseCode> {
+        if fan_count > MAX_FANS {
+            ringbuf_entry!(Trace::BadFanCount(fan_count));
+            return Err(ResponseCode::BadArg);
+        }
+
+        assert!(fan_count <= MAX_FANS);
+        let device = &self.device;
+
+        // Enable the watchdog at all times
+        let mut config =
+            Configuration(read_reg8(device, Register::Configuration)?);
+        config.set_watchdog_enable(true);
+        write_reg8(device, Register::Configuration, config.0)?;
+
+        for fan in 0..fan_count {
+            let fan = Fan::try_from(fan).unwrap();
+
+            // Configure tach stuff
+            let reg1 = fan.configuration1();
+            let mut cfg1 = FanCfg1(read_reg8(device, reg1)?);
+            cfg1.set_range(0b01); // 1000 RPM minimum, TACH count multiple = 2
+            cfg1.set_edges(0b01); // 5 edges sampled, TACH multiple = 1x
+            write_reg8(device, reg1, cfg1.0)?;
+
+            // Enable ramp control, to avoid the fans going from 0-100%
+            let reg2 = fan.configuration2();
+            let mut cfg2 = FanCfg2(read_reg8(device, reg2)?);
+            cfg2.set_enable_ramp_control(true);
+            write_reg8(device, reg2, cfg2.0)?;
+
+            write_reg8(device, fan.pwm_target(), 0)?;
+        }
+
+        Ok(())
+    }
+
+    /// Determines the rotations per minute based on the tach count
+    pub fn fan_rpm(&self, fan: Fan) -> Result<Rpm, ResponseCode> {
+        let val = read_reg16(&self.device, fan.tach_count())?;
+
+        // See section 4.4.1 in the datasheet for details
+        //
+        // The tachometer reading represents the number of 32 KHz clock cycles
+        // counted in a single revolution of the fan.
+        //
+        // So to get from the tach count to the time per revolution:
+        //
+        //              1         n - 1
+        //      RPM = ----- x ------------- x f_TACH x 60
+        //            poles    count * 1/m
+        //
+        // Our settings are the following:
+        //   poles = 2
+        //   n = edges measured (set to 5 in `cfg1` above)
+        //   m = range multiplier (2x, per configuration above)
+        //   f_TACH = tachometer measurement frequency (32 KHz)
+        //   count = reading
+        //
+        // Doing the math, we end up with
+        //   RPM = 7680000 / count
+        let count = ((val[0] as u32) << 5) | (val[1] >> 3) as u32;
```

Comment:
where do these shifts come from?  i assume this is defined in the datasheet, but it surprised me --- at a glance, i kind of assumed that we were just ORing the two bytes into a 16-bit value, but this is a bit more complex...

Reply:
Weirdly, the count is a 12-bit value that's shifted by 3.  We use 5 bits of the low byte (`val[1] >> 3`) and 8 bits of the high byte (`val[0] << 5`)